parlai/agents/bert_qa/bert_qa.py

```diff
@@ -0,0 +1,443 @@
+            batch["observations"]
+        )
+
+        with torch.no_grad():
+            b_start_logits, b_end_logits = self.model(
+                b_tokens_ids, b_segment_ids, b_input_mask
+            )
+
+        predictions = []
+        for start_logits, end_logits, input_ids, start_context_position in zip(
+            b_start_logits, b_end_logits, b_tokens_ids, b_start_context_position
+        ):
+            prediction = self._get_prediction(
+                start_logits, end_logits, input_ids, start_context_position
+            )
+            predictions.append(prediction)
+
+        return Output(predictions)
+
+    def share(self):
+        """Share model parameters."""
+        shared = super().share()
+        shared["model"] = self.model
+        shared["metrics"] = self.metrics
+        return shared
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        TorchAgent.add_cmdline_args(parser)
+        parser = parser.add_argument_group("BERT QA Arguments")
+        parser.add_argument(
+            "--n-best-size",
+            default=20,
+            type=int,
+            help="The total number of n-best predictions to generate in the nbest_predictions.json "
+            "output file.",
+        )
+        parser.add_argument(
+            "--bert-model",
+            default="bert-base-cased",
+            type=str,
+            help="Bert pre-trained model selected in the list: bert-base-uncased, "
+            "bert-large-uncased, bert-base-cased, bert-large-cased, bert-base-multilingual-uncased, "
+            "bert-base-multilingual-cased, bert-base-chinese.",
+        )
+        parser.add_argument(
+            "--warmup_proportion",
+            default=0.1,
+            type=float,
+            help="Proportion of training to perform linear learning rate warmup for. E.g., 0.1 = 10%% "
+            "of training.",
+        )
+        parser.add_argument(
+            "--loss_scale",
+            type=float,
+            default=0,
+            help="Loss scaling to improve fp16 numeric stability. Only used when fp16 set to True.\n"
+            "0 (default value): dynamic loss scaling.\n"
+            "Positive power of 2: static loss scaling value.\n",
+        )
+        parser.add_argument(
+            "--do-lower-case",
+            action="store_true",
+            help="Whether to lower case the input text. True for uncased models, False for cased models.",
+        )
+        parser.set_defaults(dict_maxexs=0)  # skip building dictionary
+
+    @staticmethod
+    def dictionary_class():
+        return BertSpanDictionaryAgent
+
+    def build_model(self):
+        self.model = BertForQuestionAnswering.from_pretrained(self.pretrained_path)
+        self.n_gpu = 0
+        if self.use_cuda:
+            self.n_gpu = torch.cuda.device_count()
+            if self.n_gpu > 0:
+                self.model = torch.nn.DataParallel(self.model)
```

Comment:
is there any reason to avoid using DistributedDataParallel?

https://github.com/facebookresearch/ParlAI/blob/27c78c4ebeb599306eccd50f9f08e4d3c20d37bd/parlai/core/torch_generator_agent.py#L410-L413